# Libraries for Neural Networks - sklearn 2 (regression)


## Loading the data

https://www.cs.toronto.edu/~delve/data/boston/bostonDetail.html

In [1]:
from sklearn.neural_network import MLPRegressor
from sklearn import datasets

In [2]:
boston = datasets.load_boston()

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (i.e.

In [3]:
inputs = boston.data
inputs

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [4]:
boston.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [5]:
outputs = boston.target
outputs

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
       35.4, 24.7, 31.6, 23.3, 19.6, 18.7, 16. , 22.2, 25. , 33. , 23.5,
       19.4, 22. , 17.4, 20.9, 24.2, 21.7, 22.8, 23.4, 24.1, 21.4, 20. ,
       20.8, 21.2, 20.3, 28. , 23.9, 24.8, 22.9, 23.9, 26.6, 22.5, 22.2,
       23.6, 28.7, 22.6, 22. , 22.9, 25. , 20.6, 28.4, 21.4, 38.7, 43.8,
       33.2, 27.5, 26.5, 18.6, 19.3, 20.1, 19.5, 19.5, 20.4, 19.8, 19.4,
       21.7, 22.8, 18.8, 18.7, 18.5, 18.3, 21.2, 19.2, 20.4, 19.3, 22. ,
       20.3, 20.5, 17.3, 18.8, 21.4, 15.7, 16.2, 18. , 14.3, 19.2, 19.6,
       23. , 18.4, 15.6, 18.1, 17.4, 17.1, 13.3, 17.8, 14. , 14.4, 13.4,
       15.6, 11.8, 13.8, 15.6, 14.6, 17.8, 15.4, 21

In [6]:
boston.target_names

AttributeError: ignored

In [7]:
inputs.shape

(506, 13)

In [8]:
outputs.shape

(506,)

## Train and test datasets

In [9]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(inputs, outputs, test_size = 0.2)

In [10]:
X_train.shape

(404, 13)

In [11]:
y_train.shape

(404,)

In [12]:
X_test.shape

(102, 13)

In [13]:
y_test.shape

(102,)

In [14]:
max(X_train[0])

377.56

In [15]:
from sklearn.preprocessing import MinMaxScaler
scaler1 = MinMaxScaler()
X_train = scaler1.fit_transform(X_train)
X_test = scaler1.transform(X_test)

In [16]:
max(X_train[0])

0.9512330425134904

In [17]:
y_train = y_train.reshape(-1, 1)
y_test = y_test.reshape(-1, 1)

In [18]:
y_train.shape

(404, 1)

In [19]:
scaler2 = MinMaxScaler()
y_train = scaler2.fit_transform(y_train)
y_test = scaler2.transform(y_test)

In [20]:
max(y_train[0])

0.33333333333333337

## Neural network (training)

In [21]:
(13 + 1) / 2

7.0

In [28]:
network = MLPRegressor(max_iter=2000, verbose=True, hidden_layer_sizes=(7))
network.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1599: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Iteration 1, loss = 2.55603987
Iteration 2, loss = 2.43131664
Iteration 3, loss = 2.31278755
Iteration 4, loss = 2.20195514
Iteration 5, loss = 2.09555555
Iteration 6, loss = 1.99361961
Iteration 7, loss = 1.89579913
Iteration 8, loss = 1.80516735
Iteration 9, loss = 1.71932173
Iteration 10, loss = 1.63970234
Iteration 11, loss = 1.56459105
Iteration 12, loss = 1.49255338
Iteration 13, loss = 1.42392691
Iteration 14, loss = 1.36093708
Iteration 15, loss = 1.30183779
Iteration 16, loss = 1.24655034
Iteration 17, loss = 1.19396668
Iteration 18, loss = 1.14478058
Iteration 19, loss = 1.09782039
Iteration 20, loss = 1.05347958
Iteration 21, loss = 1.01144662
Iteration 22, loss = 0.97136881
Iteration 23, loss = 0.93405125
Iteration 24, loss = 0.89967842
Iteration 25, loss = 0.86698512
Iteration 26, loss = 0.83503122
Iteration 27, loss = 0.80459152
Iteration 28, loss = 0.77540186
Iteration 29, loss = 0.74703312
Iteration 30, loss = 0.72006431
Iteration 31, loss = 0.69497527
Iteration 32, los

MLPRegressor(hidden_layer_sizes=7, max_iter=2000, verbose=True)

In [29]:
network.coefs_

[array([[-4.12191075e-01,  6.90073259e-01,  2.84973200e-03,
         -1.16962887e-02, -1.39752469e-01, -1.30119159e-01,
         -5.06157559e-01],
        [ 1.39292051e-01,  5.09952509e-01, -2.43896567e-01,
         -4.66039370e-02, -5.57841479e-01,  1.47737460e-01,
          2.40631096e-01],
        [ 4.27265366e-01,  1.14758662e-01, -2.81234168e-01,
          3.86016157e-01, -2.21041879e-01,  2.67034850e-01,
          3.94228461e-01],
        [ 1.55656863e-01,  9.05031643e-02, -3.10644542e-01,
         -2.30759421e-01, -2.75507774e-02, -4.89446937e-03,
         -5.67186376e-01],
        [-1.40512890e-01,  4.96407927e-01,  3.21143351e-01,
         -6.52810634e-01, -2.87328103e-01,  2.90147873e-01,
         -1.43596709e-01],
        [ 2.55586567e-01,  8.02396191e-01,  2.89912665e-01,
         -4.71282422e-01,  4.05284198e-01, -3.65072091e-01,
         -5.57046172e-01],
        [ 3.02048152e-01, -2.21891748e-02, -3.03121790e-01,
         -2.90861668e-01,  1.61580246e-01, -5.49244341e-01

In [30]:
network.intercepts_

[array([ 0.35079406, -0.03557942, -0.14498708,  0.43310747,  0.09060329,
         0.34950131, -0.04645506]), array([-0.18518457])]

In [31]:
network.n_layers_

3

In [32]:
network.n_outputs_

1

In [33]:
network.out_activation_

'identity'

## Neural network (evaluation)

In [34]:
X_test.shape

(102, 13)

In [35]:
predictions = network.predict(X_test)
predictions

array([ 0.4009316 ,  0.25068612,  0.45765436,  0.41963534,  0.43690621,
        0.33342135,  0.38235268,  0.41348257,  0.35817819,  0.23682393,
        0.68236242,  0.22118387,  0.21646371,  0.38358741,  0.33085116,
        0.41244317,  0.19195457,  0.3353893 ,  0.24109654,  0.32148551,
        0.19668561,  0.4346512 ,  0.28129253,  0.34469913,  0.19220728,
        0.32785226,  0.75185419,  0.35654959,  0.36621248,  0.63768053,
        0.5317879 ,  0.52130633,  0.5016519 ,  0.17012865, -0.16363281,
        0.3660389 ,  0.1421984 ,  0.36322068,  0.31015263,  0.3795091 ,
        0.40832866,  0.41998224,  0.30919522,  0.07610374,  0.50570776,
        0.46471187,  0.37757695,  0.40270334,  0.50937493,  0.34501818,
        0.5203178 ,  0.37022865,  0.3666245 ,  0.2077814 ,  0.36938658,
        0.41114279,  0.53791136,  0.24944129,  0.31422867,  0.44049844,
        0.40615078,  0.39167484,  0.35820235,  0.38275202,  0.53318672,
        0.26509686,  0.31575779,  0.24749191,  0.40723587,  0.16

In [36]:
y_test

array([[0.17777778],
       [0.61111111],
       [0.27555556],
       [0.25777778],
       [0.16666667],
       [0.40222222],
       [0.5       ],
       [0.39111111],
       [0.49111111],
       [0.39555556],
       [0.48      ],
       [0.23555556],
       [0.34444444],
       [0.17333333],
       [0.02888889],
       [0.6       ],
       [0.34666667],
       [0.42      ],
       [0.42444444],
       [0.36      ],
       [0.41777778],
       [0.26      ],
       [0.39333333],
       [0.57111111],
       [0.19555556],
       [0.31777778],
       [0.62      ],
       [0.37111111],
       [0.33555556],
       [0.86666667],
       [0.43333333],
       [0.32      ],
       [0.19777778],
       [0.39111111],
       [0.11555556],
       [1.        ],
       [0.13333333],
       [0.62666667],
       [0.34666667],
       [0.32888889],
       [0.35333333],
       [0.23555556],
       [0.31777778],
       [0.24666667],
       [0.52222222],
       [0.50888889],
       [0.25555556],
       [0.197

In [37]:
from sklearn.metrics import mean_absolute_error, mean_squared_error
mean_absolute_error(y_test, predictions)

0.15108543864862908

In [38]:
mean_squared_error(y_test, predictions)

0.037096540463718665

In [39]:
import numpy as np
np.sqrt(mean_squared_error(y_test, predictions))

0.1926046221244928

## Neural network (classification)

In [40]:
X_test[0], y_test[0]

(array([0.0845195 , 0.        , 0.64662757, 0.        , 0.67219917,
        0.54723127, 0.98249228, 0.0959725 , 1.        , 0.91412214,
        0.80851064, 0.76627667, 0.48509934]), array([0.17777778]))

In [41]:
X_test[0].shape

(13,)

In [42]:
new = X_test[0].reshape(1, -1)
new.shape

(1, 13)

In [43]:
network.predict(new)

array([0.4009316])

In [44]:
scaler2.inverse_transform(network.predict(new).reshape(-1,1))

array([[23.04192185]])

In [45]:
scaler2.inverse_transform(y_test[0].reshape(-1,1))

array([[13.]])

** Conclusion: the results after running this neural network regression model are not very precise, the closer to 13., the better the results. We've received 23.04192185 **